### ISSUE TITLE
Using on BalenaOS

### ISSUE THREAD

**Issue by hsghost** [v2rayA/v2rayA#226]
Would like to use v2rayA on  BalenaOS to penetrate the GFW. Yet, bind mounting `/lib/modules` and `/etc/reslov.conf` form the host are not allowed. Any consequences? Transparent proxy not supported, then?

If so, any suggestions on this topic? I've tried redirecting everything from host to the v2rayA container through SOCKS5, yet the device went directly offline. I'd became a circle, I think. Novice to network protocols, please help!

Thanks,

**Comment by MarksonHon** [comment]
@hsghost If you have a SBC then you should try Arch Linux ARM, Manjaro ARM or Armbian at first, these OSs have native packages for installation.

**Comment by hsghost** [comment]
Seemingly got it to work by utilizing [17mon/china_ip_list](https://github.com/17mon/china_ip_list) on the no_proxy entries of redsocks. Now the connection quality relies solely on the reliability of the relay service I'm subscribing to. Anyway it's a more affordable solution, in comparison with registering with the IPC. Good work, thanks!

One question left though. There are entries in the `config.json` to set the listening IPs as well as ports. Yet, when I tried to change the listening address of SOCKS5 to my host IP, it got automatically reset to 0.0.0.0 once I restarted the service. So, is it a feasible option to set this listening IP to a certain address? I feel it more secure as, under the 0.0.0.0 setting, any incoming SOCKS5 request can be served by the proxy service, creating a somewhat loophole.

**Comment by mzz2017** [comment]
@hsghost If you turn "Intranet Sharing" off, it will bind 127.0.0.1. If you want to bind other host instead, close those port at "Settings"-"Address and Ports". And set up new inbounds using RoutingA.

**Comment by hsghost** [comment]
So, how should I utilize RoutingA? I've read the whole wiki page about it yet to find where to put the markdown code and where to command RoutingA to generate the V2Ray config it suppose to output. Any hints?

**Comment by mzz2017** [comment]
@hsghost Setting-prevent DNS pollution

**Comment by hsghost** [comment]
I've tested using RoutingA. If I want to disable `Share In LAN` and use RoutingA to restrict the inbound IPs, I found that it is only available when `Transparent Proxy` is enabled. Yet, it is infeasible under the current BalenaOS settings, as full privileged mode cannot be achieved. So I had to go another route - to setup an auth. I set up the following in RoutingA:

```
inbound:socksauthin=socks(address: 0.0.0.0, port: 20170, user: someuser, pass: somepass)
inboundTag(socks)->block
default: proxy
domain(geosite:geolocation-!cn)->proxy
domain(geosite:google-scholar)->proxy
domain(geosite:category-scholar-!cn, geosite:category-scholar-cn, domain:qq.com)->direct
domain(geosite:cn)->direct
ip(geoip:private, geoip:cn)->direct
```
Basically it's just `Proxy Except CN Sites` with a customized inbound. And in host redsocks I configured: 

```
redsocks {
	local_ip = 127.0.0.1;
	local_port = 12345;
	type = socks5;
	ip = 127.0.0.1;
	port = 20170;
	login = "someuser";
	password = "somepass";
}
```
But it seems to be unsuccessful. I suspect that the authorization did not go through. As a newbie, could you kindly point me some tool to track through the traffic, so that I can confirm whether they're actually going through `ascksauthin` endpoint rather than the built in `socks`? Thank you.

**Comment by mzz2017** [comment]
@hsghost Close the port in Setting - Address and Ports first

**Comment by hsghost** [comment]
Bang! What magic you casted to read exactly what detail I'd missed! Just turned the SOCKS5 port to `0` in the settings and it suddenly works like a charm! Thanks so much for all the help!

Just one curiosity question now. BalenaOS officially support the mechanism to communicate with the host OS through D-Bus. So, is it possible that we can utilize this to control the host components such as `systemd` and `NetworkManager`, etc., so that the requirements to access the host paths `/lib/modules` and `/etc/reslov.conf` can be lifted?

**Comment by mzz2017** [comment]
@hsghost Good suggest, what about other OS? It is a piece of good news if the mechanism is universally applicable for all Linux distributions.

**Comment by wulfy23** [comment]
fyi package recently added to openwrt master/snapshot...

**Comment by hsghost** [comment]
@mzz2017 As per [Wikipedia: D-Bus - History and adoption](https://en.wikipedia.org/wiki/D-Bus#History_and_adoption), D-Bus exists on most Linux distributions. It is extensively used by `systemd` and DE applications. So Ubuntu Desktop comes with it preinstalled, while other non-DE distributions like [Debian](https://packages.debian.org/search?keywords=dbus) and [Alpine](https://pkgs.alpinelinux.org/package/edge/main/x86/dbus) can be installed through their respective package managers. There are even ports to [macOS](https://formulae.brew.sh/formula/dbus) and [Windows](https://www.freedesktop.org/wiki/Software/dbus/#download), yet I doubt these ports are of any use as there's no facilities like `systemd` on these host platforms.

On the other hand, communication through D-Bus with the host OS from within a container [can be achieved](https://unix.stackexchange.com/questions/546726/setup-docker-container-to-communicate-with-host-over-d-bus). And on BalenaOS it is being pre-configured for an easier adoption: [Balena Documentation: Communicate outside the container - D-Bus communication with host OS](https://www.balena.io/docs/learn/develop/runtime/#d-bus-communication-with-host-os).

And as per @wulfy23 mentioned, an [OpenWRT package](https://openwrt.org/packages/pkgdata/dbus) is also available.

**Comment by hsghost** [comment]
Is there a chance for RoutingA to support expanding environment variables in the rules? That way I could utilize the BalenaCloud dashboard to configure the host and v2rayA end at one single place.

**Comment by mzz2017** [comment]
@hsghost It is theoretically feasible, but the variables can only be compiled at the moment v2ray-core is started

**Comment by hsghost** [comment]
@mzz2017 It is sufficient if the environment variables are being taken only at program start. The BalenaOS would restart a container instance whenever a modification is made to the variables associate to it, anyway.

So, is this an existing feature in RoutingA, or that I have to wait for the next release to try it?

**Comment by mzz2017** [comment]
@hsghost The feature hasn't been added to v2rayA yet. It is on the todolist now.

**Comment by hsghost** [comment]
@mzz2017 Okay, hard coding it for now, looking forward to the updates. Thanks!

**Comment by mzz2017** [comment]
https://github.com/v2rayA/v2rayA/actions/runs/1277452733
hard code usage:
```
export SINGLE_LINE="ip(0.0.0.0/0)->proxy"
export IP="1.1.1.1"
export OUTBOUND="block"
```
```
# RoutingA
ip($$IP$$) -> $$OUTBOUND$$
$$SINGLE_LINE$$
```

**Comment by mzz2017** [comment]
This issue will be closed because it has been fixed in v1.5.4.
